doc/dev/adr/0020-large-message-chunking.md

```diff
@@ -0,0 +1,100 @@
+- The overhead size will be configurable, large enough to simplify calculations while ensuring we stay under the broker's limit.
+
+**Implementation layer**:
+- Chunking will be implemented as middleware in the Protocol layer between serialization and MQTT client.
+ ```
+ Application → Protocol Layer (Serialization) → Chunking Middleware → MQTT Client → Broker
+ ```
+- This makes chunking transparent to application code and compatible with all serialization formats.
+
+- Sending Process:
+  - When a payload exceeds the maximum packet size, the middleware intercepts it before transmission
+  - The message is split into fixed-size chunks (with potentially smaller last chunk)
+  - Each chunk is sent as a separate MQTT message with the same topic but with chunk metadata.
+  - Effort should be made to minimize user properties copied over to every chunk: first chunk will have full set of original user properties and the rest only thoses that are neccessary to reassamble original message (ex.: ```$partition``` property to support shared subscriptions:).
+  - QoS settings are maintained across all chunks.
+- Receiving Process:
+  - The Chunking aware client receives messages and identifies chunked messages by the presence of chunk metadata.
+  - Chunks are stored in a temporary buffer, indexed by message ID and chunk index.
+  - When all chunks for a message ID are received, they are reassembled in order and message checksum verified.
+  - The reconstructed message is then processed as a single message by the application.
+
+### Benefits
+- **Property Preservation:** Maintains topic, QoS, and other message properties consistently
+- **Network Optimized:** Allows efficient transmission of large payloads over constrained networks
+
+### Implementation Considerations
+- **Error Handling:**
+  - Chunk timeout mechanisms (see Chunk Timeout Mechanism Options in the Appendix)
+  - Error propagation to application code
+- **Performance Optimization:**
+  - Concurrent chunk transmission
+  - Efficient memory usage during reassembly
+- **Security:**
+  - Validate message integrity across chunks and prevent chunk injection attacks (see Checksum Algorithm Options for MQTT Message Chunking in the Appendix)
+
+# Appendix
+
+## Chunk Timeout Mechanism Options
+
+1. Fixed Timeout Window
+   - Set a single timeout period after receiving the first chunk
+   - If all chunks aren't received within this window, the message is considered failed
+   - **Pros**: Simple implementation, predictable behavior
+   - **Cons**: Not adaptive to message size or network conditions
+
+2. Sliding Timeout Window
+   - Reset the timeout each time a new chunk arrives
+   - Only expire the chunked message if there's a long gap between chunks
+   - **Pros**: Tolerates varying network conditions and delivery rates
+   - **Cons**: Could keep resources allocated for extended periods
+
+## Checksum Algorithm Options for MQTT Message Chunking
+
+1. MD5
```

Comment:
We could consider asking users to inject their algorithm of choice so that our SDK doesn't choose one for them